architecture: add document on ztunnel lifecycle (#51805)
istio/istio: architecture/ambient/lifecycle.md

> # Ztunnel Lifecyle

Reviewer: IMO we should do one of: 

- put this in `ztunnel.md`
- name it `ztunnel-lifecycle.md`
- name it `traffic-capture-lifecycle.md`

Reviewer: Also we have:

`TODO: fill in implementation details of how redirection is actually implemented.`

in `ztunnel.md` -> that TODO should probably just link to this doc.

Reviewer: Also - we should probably _remove_ some of the text in https://github.com/istio/istio/tree/master/cni/README.MD in favor of this - there's some duplication, and the former is slightly out of date.

This document describes the lifecycle of various aspects of Ztunnel, and its relationship to workloads and the CNI.

At a high level, our goal is to provide complete connectivity to a workload throughout its entire lifetime.
Failing to do so can be an availability risk (if we deny traffic that should succeed) or a security risk (if we allow traffic that should be denied).

## High level overview

At a high level, the relevant components look as such:

```mermaid
flowchart TD
    CNIP["CNI Plugin"]
    CNI["CNI Agent"]
    Z["Ztunnel"]
    Pods["Pods"]
    CNI --Programs--> Z
    CNI --Watches--> Pods
    Z --Runs Within--> Pods
    CNIP --Calls--> CNI
```

> The CNI Plugin is a binary installed as a [CNI plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) on the node.

Reviewer: Should we just put some/all of this stuff in https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ ?

I feel like most of the generic `lifecycle` stuff and calling out of specific security considerations and our mitigations of them would be useful to have there.

e.g. "It is security-critical to handle pod startup gaps for this reason, and we do in this way" probably belongs in `istio.io` docs.

I don't want to document the CNI node agent in too many disparate places, we already have

- https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ (ambient-specific generic traffic capture architecture)

- https://istio.io/latest/docs/setup/additional-setup/cni/ (dataplane-mode agnostic install guide)
  
cc @craigbox for thoughts.

Author: I agree some probably should be in public docs but there is serious value to having developer docs like this that is not tied up in public docs personas

Reviewer 2: I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

What is relevant for public docs is "The ambient L4 layer guarantees that network is secure ... from start to end"

Reviewer: Implementation details should not be covered in public docs - agreed.

But I don't think "K8S-specific pod lifecycle security considerations" are component-level implementation details. That's the whole reason/justification we use a CNI plugin, and have repair for sidecars, etc etc. It's worth explaining that at a high level.

We have already had people ask about these basic architectural justifications, is why I'm suggesting we have something in public docs.

Maybe what's already there is sufficient - that's fine. Wanna make sure it is.

Reviewer: > I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

I don't think we can get away with that, since the entire reason we have a node agent with a CNI plugin has to do with K8S pod lifecycle implementation details.

Author: Implementation details in docs are good for a few reasons:
* You did something great and need to prove it's real (we provide secure networking with X properties,and here is the details if you really want)
* You did something terrible and need to explain it (sidecar init has issues so we hacked on a repair controller)
* Understanding details is useful for operations (arguably, this is the same as "you did something terrible", because making users care about low level details means we did a bad job making it operational)

I think we have a bit of each here (hopefully less terrible stuff)

Reviewer: Yeah - I'm more concerned that we have those implementation details here _and_ elsewhere too, and I'd rather be parsimonious about duplicating those details in different spots.

Putting them here is fine - if we do a pass to remove those same details from the existing spots, if and where they exist.

Reviewer 3: I agree that this is content for two separate audiences and it's OK to have some overlap.
Once this is in, we can easily mine the best bits and see what should go into the user-facing documentation.

Reviewer: Then let's at least link to this from the `istio-cni` README then, since it already has a TODO stub for it.

Reviewer: (and scrub a lot of the content from that README too, actually - it mostly restates a subset of what's here now)

> Kubernetes is responsible for invoking this when a Pod is being started (before the containers run).

Reviewer 2: Technically Ambient is not restricted to K8S - I would say 'the execution environment' - unless you change the title to "Ztunnel per-node daemon set" lifecycle ( which is probably better, lifecycle on a VM or as a container are different enough ).

Author: yeah I think this is "ztunnel as a daemon set" architecture. I'll update

> Kubernetes is responsible for invoking this when a Pod is being started (before the containers run).

Reviewer: ```suggestion
The container runtime is responsible for invoking this when a Pod is being started (before the containers run).
```

Reviewer 3: still holding out hope for Mesos I see

Reviewer: obiwan.gif

> When this occurs, the plugin will call out to the CNI Agent to program the network.

Reviewer 2: K8S will call the Istio CNI binary (path = ??? ) , which will use a UDS socket ( path = ???) to communicate the pod info to the server running in the Istio CNI Daemon. 

The Istio CNI daemon will keep an in-memory cache with all the pods and their corresponding network namespaces.

> This includes setting up networking rules both within the pod network namespace, and also on the host network.

Reviewer 4: grammar nit:
```suggestion
This includes setting up networking rules within both the pod network namespace and the host network.
```

> This includes setting up networking rules both within the pod network namespace, and also on the host network.

Reviewer 2: Commenting as I read - do we have some examples with what is set on the host network - and what is set on the pod network ? The first part in particular - since it's different from normal Istio where we don't set anything on the host.

Reviewer: Some of that (should be/was) in https://github.com/istio/istio/tree/master/cni/README.MD

All that is created on the host are a single iptables rule and ipset for host probes.

That does raise a good question - if this is `/architecture/ambient` - should it have that impl detail in it?

Or should that impl detail go in `cni/README.MD`, next to the thing that actually does the impl?

Author: I linked to CNI/readme.dm

> An alternative flow is when a pod is enabled after startup.

Reviewer 2: Enabled after which startup ? How can that happen ( the CNI binary will be always called).
You mean 'if the CNI binary is installed later ' ?

Author: Its if the namespace/pod is enrolled after it starts. By label, etc

In this case, the CNI Agent is watching for Pod events from the API server directly and performing the same setup.

> 

Reviewer 2: This deserves a lot more details - for developers and admins - it is not even close to the same setup. 

Watching the pods doesn't provide the network namespace - we perform a duct-taped search in /proc with some regex to guess which processes are pods and what are their namespaces, hoping the CRI layer is one that happens to generate ( and guarantee) a particular naming pattern that was reverse engineered from undocumented and unsupported behaviors of common CRIs.

I think it is critical for any dev ( or operator ) to understand this part - it's one of the most tricky areas.

Once the network is configured, the CNI Agent will signal to Ztunnel to start running within the Pod.
This is done by the [ZDS](../../pkg/zdsapi/zds.proto) API.
This will send some identifying information about the Pod to Ztunnel, and, importantly, the Pod's network namespace file descriptor.

Ztunnel will use this to enter the Pod network namespace and start various listeners (inbound, outbound, etc).

> [!NOTE]
> While Ztunnel runs as a single shared binary on the node, each individual pod gets its own unique set of listeners within its own network namespace.

## Pod Startup

### Pod Startup Requirements

> During Pod startup, we MUST meet the constraint that the network is ready the instant the application starts (including `initContainers`!).

Reviewer: ```suggestion
During new Pod startup, we MUST meet the constraint that the network is ready the instant the application starts (including `initContainers`!).
```

This is critical -- issues around this were the [top issue with sidecars](https://github.com/istio/istio/issues/11130).

### Pod Startup Implementation

The key mechanism to maintain this property is the CNI plugin.

> The CNI plugin synchronously runs before application containers start, ensure we can setup the network configuration and start Ztunnel.

Reviewer: ```suggestion
The CNI plugin is executed synchronously during the creation of the Pod sandbox by the container runtime, before processes or containers are scheduled in the Pod. This ensures we can setup the network configuration and start Ztunnel.
```

> Only once those two conditions are met do we mark the CNI plugin as successfully complete.

Reviewer 2: What are the error cases ? What happens if ztunnel pod is not deployed ? 

My understanding is that CNI pod will keep in memory a list with all the network namespaces, and send them to ztunnel if it shows up ( and if CNI daemon is restarted - will do the /proc scanning to guess )

Author: It blocks the pod startup. I updated the doc

> Only once those two conditions are met do we mark the CNI plugin as successfully complete.

Reviewer: ```suggestion
Only once those two conditions are met does the CNI plugin return success - effectively allowing our plugin to block Pod scheduling until we are ready.

If the CNI plugin does not successfully complete, the Pod sandbox will fail to complete, and the Pod will not be scheduled by Kubernetes.
```

From the above we are ensured that when the application starts and opens a connection, Ztunnel will be ready to `accept()` it.
However, this is not enough to fully serve the request - Ztunnel will need the certificate (from the CA) and full workload information (from the XDS server).
This is done through a waiting mechanism: Ztunnel will accept all connections, but it will not start forwarding traffic until all the above is fetched.
This "lazy loading" optimizes pod startup time, for the common case where an application is not sending outbound traffic instantly.
For applications that do, they will see a slight latency increase during startup; most applications can handle this far better than failed connections.

## Pod Shutdown

### Pod Shutdown Requirements

> During Pod shutdown, we MUST meet the constraint that the Pod shutting down can send traffic up until its full stopped.

Reviewer: ```suggestion
During Pod shutdown, we MUST meet the constraint that Pod traffic will be handled until all containerized processes within the Pod have terminated. We do this by waiting for K8S to tell us that the Pod has terminated, as it is the final/best/last arbiter for this.
```

Pod shutdown can be a long process; while `terminationGracePeriodSeconds` is typically ~30s, it can be extremely high, leaving terminating pods running for extended periods of time with an expectation of networking.

> Additionally, we MUST accept incoming traffic during this time, both new and existing.

Reviewer: Most of this can be rolled into the statement "we don't remove until we get a pod terminated state from K8S, K8S is the arbiter"

> Typically, it is rare to get new connections after shutdown has started - once shutdown is triggered the endpoint is marked `NotReady` and removed from load balancing consideration.

Reviewer 2: For example using persistent sessions or pod-to-pod communication will keep sending new connections after shutdown has started.

However, some applications do not utilize Service readiness (connecting directly to the pod, etc).
Additionally, eventually consistency means we will often get some final requests trickle in before all clients are aware the Pod is shutting down.

Finally, we SHOULD tell our (HBONE) peers we are shutting down.
This can help ensure that clients do not hold onto pooled connections for too long.
This is not strictly required; Ztunnel clients have both keepalives (which will start to fail once the other end is shut down) and pool idle timeouts (which will close the connection when it has had no activity for some period; because the backend is gone, there will be no activity).
However, it would be ideal to give immediate feedback rather than relying on these mechanisms.

### Pod Shutdown Implementation

[ZDS](../../pkg/zdsapi/zds.proto) exposes a `DelWorkload` message to signal to Ztunnel that a pod has been terminated.
This can be triggered either by a Pod being actually deleted, or has transitioned to a [terminal phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/).
At this point, the pod and all containers have fully shut down, and we can tear everything down.

> Because the application(s) are already entirely gone, there is no need to have any "draining" time -- we can immediately terminate.

Author: Note: there is no need, but at the time of writing we do drain. I willwork on fixing it.

Reviewer: xref: https://github.com/istio/ztunnel/issues/899

Furthermore, because the Pod is torn down, we cannot even send any data at all (at least on most clusters, the `veth` is torn down upon Pod deletion).

> Notably, this prevents us from sending a GOAWAY to notify peers we have shutdown.

Reviewer: ```suggestion
Notably, this prevents us from sending a GOAWAY to notify peers we have shutdown, even in the cases where we could in theory do so (i.e. clients understand and respect GOAWAY)
```

As a result, while we meet our two must-haves (send and accept traffic throughout the entire pod lifetime), we do not have an optimal solution here yet.

> Perhaps hooking into the [CNI `DEL`](https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications) could be a path forward here.

Reviewer: ```suggestion
```

This probably just belongs in an issue where the problem can be discussed, not this doc - the hook (or lack thereof) isn't really the problem here AFAIK, it's that you can't GOAWAY clients that aren't aware they're being proxied.

Author: No, that is not really the problem I am trying to convey. The problem is HBONE clients, which we know can GOAWAY, cannot because we cannot send it at the right time. This isn't talking about the fact we are handling opaque TCP streams.

I will remove this CNI DEL part though, agree with your comment

## Ztunnel Shutdown/Upgrade/Restart

This could be upgrading to a new version, or changing some part of the DaemonSet spec, etc.

### Ztunnel Shutdown Requirements

Our high level requirement is to minimize traffic disruption.

However, because Ztunnel operates only at Layer 4, we are in a tricky position:
* TCP is stateful, so we cannot really just pass state over to the other process.
  If we operated at L3 we could just start processing packets in the new Ztunnel.
* We do not operate at L7, so we cannot signal to the application, in any way, that it should reconnect (to the new Ztunnel).

Realistically, that means the best we can do is:
* Ensure that, at any point in time, any _new_ connections will succeed. There is no period where new connections are dropped.
* Provide some period of time for the old Ztunnel to continue processing any established connections.

If this period of time ("drain period") is longer than any connections, there is zero impact on applications at all.
If not, however, at some point the TCP connection will be torn down.

> Depending on the application, this may have a range of impacts (terminating in-flight transactions, etc).

Reviewer: ```suggestion
Depending on the application, this may have a range of impacts (terminating in-flight transactions, etc). In general, applications should never assume the network is 100% reliable and that no connections should be disrupted, or they will generally have[ stability problems in the context of any distributed system](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing)
```

### Ztunnel Shutdown Implementation

> The key to this "handoff" between Ztunnels is `SO_REUSEPORT`, which Ztunnel sets on all of its listeners.

Reviewer: ```suggestion
The key to this "handoff" between Ztunnels is `SO_REUSEPORT`, which Ztunnel by default sets on all of its listeners.
```

> This enables multiple processes to bind to the same port.

Reviewer: ```suggestion
This enables multiple processes with the same effective UID to bind to the same port.
```

> Linux will then send new connections to one of the processes.

Reviewer: ```suggestion
Linux will then send new connections to one of the processes.
```

Which one? Does Linux actually do this, or is it just a best-effort case of "first-acceptor-wins"?

This still requires delicate sequencing.
Consider a rolling restart of the DaemonSet.
We will do the following:

1. `ztunnel-new` starts, connects to CNI.
1. CNI sends the current state of all pods on the node. Ztunnel establishes listeners in each pod running on the node, and marks itself "ready".

> 1. At this point, we have both Ztunnels listening. New connections will be assigned to _either one_.

Reviewer: This isn't a true statement in all cases, it's highly dependent on your rolling update strategy and how you configured the K8S daemonset in general - worth pointing out, even though we do hardcode that currently.

It is, strictly speaking, something that K8S orchestrates, not us.
